File: ia/minesweeper_ai.php
<?php
// minesweeper_ai.php

require __DIR__ . '/../vendor/autoload.php';
use Ratchet\Client\WebSocket;
use Ratchet\Client\Connector;
use React\EventLoop\Factory;

$aiLevels = [
    'easy' => 0.7,  // 70% chance to make the correct move
    'medium' => 0.85, // 85% chance to make the correct move
    'hard' => 0.95  // 95% chance to make the correct move
];

// Load AI accounts
$iaAccounts = json_decode(file_get_contents('ia_accounts.json'), true);

$loop = Factory::create();
$connector = new Connector($loop);

// Variables de configuration
$inviteAutomatically = in_array('--invite', $argv);  // Si '--invite' est passé comme argument, l'IA invite elle-même
$pauseDuration = 1; // Pause par défaut en millisecondes
$invitedPlayerId = null; // Stocke l'ID du joueur invité pour réinviter plus tard

// Vérification des arguments de lancement pour la pause
foreach ($argv as $arg) {
    if (preg_match('/--pause=(\d+)/', $arg, $matches)) {
        $pauseDuration = (int)$matches[1];  // Extraire la durée de la pause
    }
}

// AI configuration
$selectedLevel = 'medium'; // Can be 'easy', 'medium', or 'hard'
$accuracy = $aiLevels[$selectedLevel]; // The probability that the AI makes the correct move

// Get an available AI account
$aiAccount = $iaAccounts[array_rand($iaAccounts)];
$username = $aiAccount['username'];
$password = $aiAccount['password'];

// Variables pour stocker l'état du jeu
$currentGameId = null;
$currentPlayerId = null;
$currentInvitationId = null;

$connector('ws://localhost:8080')
    ->then(function (WebSocket $conn) use ($loop, $username, $password, $accuracy, $inviteAutomatically, $pauseDuration, &$currentGameId, &$currentPlayerId, &$currentInvitationId, &$invitedPlayerId) {
        // Login to the server
        $conn->send(json_encode([
            'type' => 'login',
            'username' => $username,
            'password' => $password
        ]));

        $conn->on('message', function ($msg) use ($conn, $accuracy, $username, $inviteAutomatically, $pauseDuration, &$currentGameId, &$currentPlayerId, &$currentInvitationId, &$invitedPlayerId) {
            $data = json_decode($msg, true);

            switch ($data['type']) {
                case 'login_success':
                    echo "Logged in as AI: " . $data['username'] . PHP_EOL;
                    $currentPlayerId = $data['playerId']; // Stocke l'ID du joueur IA

                    // Si l'invitation automatique est activée, chercher un joueur à inviter
                    if ($inviteAutomatically) {
                        searchAndInvitePlayer($conn, $data['players'], $invitedPlayerId);
                    }
                    break;

                case 'invite':
                    // Automatically accept the invitation
                    echo "Invitation received from: " . $data['inviter'] . PHP_EOL;
                    $currentInvitationId = $data['invitationId']; // Stocker l'invitation
                    $conn->send(json_encode([
                        'type' => 'accept_invite',
                        'invitationId' => $currentInvitationId
                    ]));
                    break;

                case 'game_start':
                    // Game starts, store game_id and process the board like in update_board
                    echo "Game started!" . PHP_EOL;
                    $currentGameId = $data['game_id']; // Stocker l'ID de la partie
                    displayBoard($data['board']);
                    makeMove($conn, $data['board'], $currentGameId, $accuracy, $pauseDuration);
                    break;

                case 'update_board':
                    // Check if it's the AI's turn
                    if ($data['currentPlayer'] === $username) {
                        echo "It's my turn!" . PHP_EOL;
                        displayBoard($data['board']);
                        makeMove($conn, $data['board'], $currentGameId, $accuracy, $pauseDuration);
                    } else {
                        echo "Waiting for my turn..." . PHP_EOL;
                        displayBoard($data['board']);
                    }
                    break;

                case 'game_over':
                    echo "Game over. Winner: " . $data['winner'] . PHP_EOL;
                    displayBoard($data['board']);
                    
                    // Réinviter le même joueur après la fin de la partie
                    if ($inviteAutomatically && $invitedPlayerId !== null) {
                        echo "Reinviting player with ID: " . $invitedPlayerId . PHP_EOL;
                        invitePlayer($conn, $invitedPlayerId);
                    }
                    break;

                case 'player_disconnected':
                    echo "Opponent disconnected. The game is over." . PHP_EOL;
                    break;

                case 'connected_players':
                    // Si l'invitation automatique est activée, chercher un joueur à inviter
                    if ($inviteAutomatically) {
                        searchAndInvitePlayer($conn, $data['players'], $invitedPlayerId);
                    }
                    break;

                case 'error':
                    echo "Error: " . $data['message'] . PHP_EOL;
                    break;

                default:
                    echo "Unknown message type received: " . $data['type'] . PHP_EOL;
                    break;
            }
        });

        $conn->on('close', function ($code = null, $reason = null) use ($loop) {
            echo "Connection closed ({$code} - {$reason})\n";
            $loop->stop();
        });
    }, function ($e) use ($loop) {
        echo "Could not connect: {$e->getMessage()}\n";
        $loop->stop();
    });

$loop->run();

// Fonction pour afficher le plateau de jeu
function displayBoard($board) {
    $width = count($board);
    $height = count($board[0]);

    echo "Current board state:\n";
    for ($x = 0; $x < $width; $x++) {
        for ($y = 0; $y < $height; $y++) {
            $cell = $board[$x][$y];
            if ($cell['revealed']) {
                if (isset($cell['mine']) && $cell['mine']) {
                    echo "💣 "; // Mine
                } elseif ($cell['adjacentMines'] > 0) {
                    echo $cell['adjacentMines'] . " "; // Nombre de mines adjacentes
                } else {
                    echo "  "; // Case vide
                }
            } elseif ($cell['flagged']) {
                echo "🚩 "; // Drapeau
            } else {
                echo "⬜ "; // Case non révélée
            }
        }
        echo "\n";
    }
    echo "\n";
}

// Fonction pour rechercher un joueur dont le nom commence par 'ia_' et envoyer une invitation
function searchAndInvitePlayer($conn, $players, &$invitedPlayerId) {
    foreach ($players as $player) {
        if (strpos($player['username'], 'ia_') === 0 && $player['username'] !== $GLOBALS['username']) {
            echo "Inviting player: " . $player['username'] . PHP_EOL;
            // Envoyer l'invitation avec une grille 10x10 et difficulté facile (15%)
            invitePlayer($conn, $player['id']);
            $invitedPlayerId = $player['id']; // Stocker l'ID du joueur invité
            return; // Arrêter après avoir invité un joueur
        }
    }
    echo "No suitable 'ia_' player found to invite." . PHP_EOL;
}

// Fonction pour inviter un joueur
function invitePlayer($conn, $playerId) {
    $conn->send(json_encode([
        'type' => 'invite',
        'invitee' => $playerId,
        'gridSize' => '10x10',   // Taille par défaut de la grille
        'difficulty' => 15       // Difficulté par défaut
    ]));
}

// Fonction pour effectuer un coup basé sur l'état du plateau
function makeMove($conn, $board, $currentGameId, $accuracy, $pauseDuration) {
    $width = count($board);
    $height = count($board[0]);
    
    $safeMoves = [];
    $possibleMoves = [];

    // Collecter toutes les cases non révélées et analyser les cases révélées avec des mines adjacentes
    for ($x = 0; $x < $width; $x++) {
        for ($y = $y = 0; $y < $height; $y++) {
            if (!$board[$x][$y]['revealed'] && !$board[$x][$y]['flagged']) {
                // Case non révélée et non marquée comme drapeau
                $possibleMoves[] = ['x' => $x, 'y' => $y];
            } elseif ($board[$x][$y]['revealed'] && isset($board[$x][$y]['adjacentMines']) && $board[$x][$y]['adjacentMines'] > 0) {
                // Case révélée avec des mines adjacentes, analyser les voisins
                $neighbors = getNeighbors($x, $y, $width, $height);
                $unrevealedNeighbors = [];
                $flaggedNeighbors = 0;

                foreach ($neighbors as $neighbor) {
                    if (!$board[$neighbor['x']][$neighbor['y']]['revealed']) {
                        if ($board[$neighbor['x']][$neighbor['y']]['flagged']) {
                            $flaggedNeighbors++;
                        } else {
                            $unrevealedNeighbors[] = $neighbor;
                        }
                    }
                }

                // Si toutes les mines adjacentes sont marquées, les autres voisins sont sûrs
                if ($flaggedNeighbors == $board[$x][$y]['adjacentMines']) {
                    $safeMoves = array_merge($safeMoves, $unrevealedNeighbors);
                }
            }
        }
    }

    // Décider quel coup jouer
    if (!empty($safeMoves)) {
        // Jouer un coup sûr
        $selectedMove = $safeMoves[array_rand($safeMoves)];
        echo "Safe move found at ({$selectedMove['x']}, {$selectedMove['y']})\n";
    } else {
        // Si aucun coup sûr, choisir au hasard
        $selectedMove = $possibleMoves[array_rand($possibleMoves)];
        echo "Random move chosen at ({$selectedMove['x']}, {$selectedMove['y']})\n";
    }

    // Pause avant d'envoyer le coup
    usleep($pauseDuration * 1000); // Pause en millisecondes

    // Envoyer le coup au serveur
    $conn->send(json_encode([
        'type' => 'reveal_cell',
        'game_id' => $currentGameId,
        'x' => $selectedMove['x'],
        'y' => $selectedMove['y']
    ]));

    echo "Move made at ({$selectedMove['x']}, {$selectedMove['y']})\n";
}

// Fonction pour obtenir les voisins valides d'une cellule
function getNeighbors($x, $y, $width, $height) {
    $neighbors = [];
    for ($dx = -1; $dx <= 1; $dx++) {
        for ($dy = -1; $dy <= 1; $dy++) {
            if ($dx === 0 && $dy === 0) {
                continue;
            }
            $nx = $x + $dx;
            $ny = $y + $dy;
            if ($nx >= 0 && $ny >= 0 && $nx < $width && $ny < $height) {
                $neighbors[] = ['x' => $nx, 'y' => $ny];
            }
        }
    }
    return $neighbors;
}

File: ia/minesweeper_ai.py
import asyncio
import websockets
import json
import random
import time
import sys

# Configuration des niveaux d'IA
ai_levels = {
    'easy': 0.7,  # 70% chance de faire un bon coup
    'medium': 0.85,  # 85% chance de faire un bon coup
    'hard': 0.95  # 95% chance de faire un bon coup
}

# Chargement des comptes IA
with open('ia_accounts.json', 'r') as f:
    ia_accounts = json.load(f)

# Variables de configuration
pause_duration = 1000  # Pause par défaut en millisecondes
invite_automatically = '--invite' in sys.argv  # Si '--invite' est passé en argument
invited_player_id = None  # Stocke l'ID du joueur invité pour réinviter plus tard

# Vérification des arguments de lancement pour la pause
for arg in sys.argv:
    if arg.startswith('--pause='):
        pause_duration = int(arg.split('=')[1])  # Extraire la durée de la pause

# Configuration IA
selected_level = 'medium'  # Peut être 'easy', 'medium', ou 'hard'
accuracy = ai_levels[selected_level]

# Variables pour stocker l'état du jeu
current_game_id = None
current_player_id = None
current_invitation_id = None


async def connect_to_server(uri):
    async with websockets.connect(uri) as websocket:
        await attempt_login(websocket)


async def attempt_login(websocket):
    """Tente de se connecter avec un compte au hasard jusqu'à succès"""
    available_accounts = ia_accounts.copy()
    while available_accounts:
        # Choisir un compte IA au hasard
        ai_account = random.choice(available_accounts)
        username = ai_account['username']
        password = ai_account['password']
        
        print(f"Trying to log in with {username}...")
        
        login_message = {
            'type': 'login',
            'username': username,
            'password': password
        }
        await websocket.send(json.dumps(login_message))
        
        # Attendre la réponse du serveur après l'envoi du login
        message = await websocket.recv()
        data = json.loads(message)
        
        if data['type'] == 'login_success':
            print(f"Logged in successfully as {username}")
            global current_player_id
            current_player_id = data['playerId']
            # Démarrer la gestion des messages après connexion
            await handle_server_messages(websocket, username)
            break
        elif data['type'] == 'login_failed':
            print(f"Login failed for {username}. Trying another account...")
            available_accounts.remove(ai_account)  # Supprimer l'IA qui a échoué


async def handle_server_messages(websocket, username):
    """Gère les messages reçus du serveur après la connexion"""
    global current_game_id, current_invitation_id, invited_player_id

    async for message in websocket:
        data = json.loads(message)

        if data['type'] == 'login_success':
            print(f"Logged in as AI: {data['username']}")
            current_player_id = data['playerId']

            # Si l'invitation automatique est activée, chercher un joueur à inviter
            if invite_automatically:
                await search_and_invite_player(websocket, data['players'], username)

        elif data['type'] == 'invite':
            print(f"Invitation received from: {data['inviter']}")
            current_invitation_id = data['invitationId']
            await accept_invite(websocket)

        elif data['type'] == 'game_start':
            print("Game started!")
            current_game_id = data['game_id']
            display_board(data['board'])
            await make_move(websocket, data['board'])

        elif data['type'] == 'update_board':
            if data['currentPlayer'] == username:
                print("It's my turn!")
                display_board(data['board'])
                await make_move(websocket, data['board'])
            else:
                print("Waiting for my turn...")
                display_board(data['board'])

        elif data['type'] == 'game_over':
            print(f"Game over. Winner: {data['winner']}")
            display_board(data['board'])

            # Réinviter le même joueur après la fin de la partie
            if invite_automatically and invited_player_id is not None:
                print(f"Reinviting player with ID: {invited_player_id}")
                await invite_player(websocket, invited_player_id)

        elif data['type'] == 'connected_players':
            if invite_automatically:
                await search_and_invite_player(websocket, data['players'], username)

        elif data['type'] == 'error':
            print(f"Error: {data['message']}")


async def accept_invite(websocket):
    """Accepter une invitation"""
    accept_message = {
        'type': 'accept_invite',
        'invitationId': current_invitation_id
    }
    await websocket.send(json.dumps(accept_message))


async def search_and_invite_player(websocket, players, username):
    """Chercher un joueur dont le nom commence par 'ia_' et envoyer une invitation"""
    global invited_player_id
    for player in players:
        if player['username'].startswith('ia_') and player['username'] != username:
            print(f"Inviting player: {player['username']}")
            await invite_player(websocket, player['id'])
            invited_player_id = player['id']
            return
    print("No suitable 'ia_' player found to invite.")


async def invite_player(websocket, player_id):
    """Inviter un joueur à jouer une partie"""
    invite_message = {
        'type': 'invite',
        'invitee': player_id,
        'gridSize': '10x10',  # Taille par défaut de la grille
        'difficulty': 15  # Difficulté par défaut
    }
    await websocket.send(json.dumps(invite_message))


async def make_move(websocket, board):
    """Effectuer un coup basé sur l'état du plateau"""
    global pause_duration
    safe_moves = []
    possible_moves = []

    width = len(board)
    height = len(board[0])

    # Collecter toutes les cases non révélées et analyser les cases révélées avec des mines adjacentes
    for x in range(width):
        for y in range(height):
            if not board[x][y]['revealed'] and not board[x][y]['flagged']:
                possible_moves.append({'x': x, 'y': y})
            elif board[x][y]['revealed'] and 'adjacentMines' in board[x][y] and board[x][y]['adjacentMines'] > 0:
                # Case révélée avec des mines adjacentes, analyser les voisins
                neighbors = get_neighbors(x, y, width, height)
                unrevealed_neighbors = []
                flagged_neighbors = 0

                for neighbor in neighbors:
                    if not board[neighbor['x']][neighbor['y']]['revealed']:
                        if board[neighbor['x']][neighbor['y']]['flagged']:
                            flagged_neighbors += 1
                        else:
                            unrevealed_neighbors.append(neighbor)

                # Si toutes les mines adjacentes sont marquées, les autres voisins sont sûrs
                if flagged_neighbors == board[x][y]['adjacentMines']:
                    safe_moves.extend(unrevealed_neighbors)

    # Décider quel coup jouer
    if safe_moves:
        selected_move = random.choice(safe_moves)
        print(f"Safe move found at ({selected_move['x']}, {selected_move['y']})")
    else:
        selected_move = random.choice(possible_moves)
        print(f"Random move chosen at ({selected_move['x']}, {selected_move['y']})")

    # Pause avant d'envoyer le coup
    time.sleep(pause_duration / 1000)  # Pause en secondes

    # Envoyer le coup au serveur
    move_message = {
        'type': 'reveal_cell',
        'game_id': current_game_id,
        'x': selected_move['x'],
        'y': selected_move['y']
    }
    await websocket.send(json.dumps(move_message))

    print(f"Move made at ({selected_move['x']}, {selected_move['y']})")


def display_board(board):
    """Affichage du plateau de jeu"""
    width = len(board)
    height = len(board[0])

    print("Current board state:")
    for x in range(width):
        for y in range(height):
            cell = board[x][y]
            if cell['revealed']:
                if 'mine' in cell and cell['mine']:
                    print("💣", end=" ")
                elif cell['adjacentMines'] > 0:
                    print(cell['adjacentMines'], end=" ")
                else:
                    print("  ", end=" ")
            elif cell['flagged']:
                print("🚩", end=" ")
            else:
                print("⬜", end=" ")
        print()


def get_neighbors(x, y, width, height):
    """Obtenir les voisins valides d'une cellule"""
    neighbors = []
    for dx in range(-1, 2):
        for dy in range(-1, 2):
            if dx == 0 and dy == 0:
                continue
            nx = x + dx
            ny = y + dy
            if 0 <= nx < width and 0 <= ny < height:
                neighbors.append({'x': nx, 'y': ny})
    return neighbors


if __name__ == "__main__":
    uri = "ws://localhost:8080"
    asyncio.get_event_loop().run_until_complete(connect_to_server(uri))
